resolve error Ranges not supported for objects of this type

jtrmal opened the pull request:
This should be resolving the following problem
```
$ /export/a09/jtrmal/kaldi-clean-git/src/featbin/apply-cmvn --norm-means=false --norm-vars=false --utt2spk=ark:data/dev10h.pem_hires_nopitch//split121/99/utt2spk scp:data/dev10h.pem_hires_nopitch//split121/99/cmvn.scp scp:data/dev10h.pem_hires_nopitch//split121/99/feats.scp ark:/dev/null
/export/a09/jtrmal/kaldi-clean-git/src/featbin/apply-cmvn --norm-means=false --norm-vars=false --utt2spk=ark:data/dev10h.pem_hires_nopitch//split121/99/utt2spk scp:data/dev10h.pem_hires_nopitch//split121/99/cmvn.scp scp:data/dev10h.pem_hires_nopitch//split121/99/feats.scp ark:/dev/null
ERROR (apply-cmvn[5.2]:ExtractObjectRange():util/kaldi-holder.h:233) Ranges not supported for objects of this type.

[ Stack-Trace: ]

kaldi::MessageLogger::HandleMessage(kaldi::LogMessageEnvelope const&, char const*)
kaldi::MessageLogger::~MessageLogger()
bool kaldi::ExtractObjectRange<kaldi::GeneralMatrix>(kaldi::GeneralMatrix const&, std::string const&, kaldi::GeneralMatrix*)
kaldi::KaldiObjectHolder<kaldi::GeneralMatrix>::ExtractRange(kaldi::KaldiObjectHolder<kaldi::GeneralMatrix> const&, std::string const&)
kaldi::SequentialTableReaderScriptImpl<kaldi::KaldiObjectHolder<kaldi::GeneralMatrix> >::EnsureObjectLoaded()
kaldi::SequentialTableReaderScriptImpl<kaldi::KaldiObjectHolder<kaldi::GeneralMatrix> >::Value()
kaldi::SequentialTableReader<kaldi::KaldiObjectHolder<kaldi::GeneralMatrix> >::Value()
main
__libc_start_main
/export/a09/jtrmal/kaldi-clean-git/src/featbin/apply-cmvn() [0x44da79]
```

danpovey commented on src/util/kaldi-holder.cc:
```diff
@@ -23,9 +23,11 @@
-bool ExtractObjectRange(const Matrix<Real> &input, const std::string &range,
-                        Matrix<Real> *output) {
+bool ParseMatrixRangeSpecifier(const std::string &range,
+                         const int rows, const int cols,
```
Now that this doesn't have a separate declaration with documentation, it needs some documentatioh.

danpovey commented on src/util/kaldi-holder.cc:
```diff
@@ -37,41 +39,104 @@ bool ExtractObjectRange(const Matrix<Real> &input, const std::string &range,
+
+bool ExtractObjectRange(const GeneralMatrix &input, const std::string &range,
+                        GeneralMatrix *output) {
+
```
redundant blank line

danpovey commented on src/util/kaldi-holder.cc:
```diff
@@ -37,41 +39,104 @@ bool ExtractObjectRange(const Matrix<Real> &input, const std::string &range,
+    const CompressedMatrix &in = input.GetCompressedMatrix();
+    ExtractObjectRange(in, range, &out);
+  } else if (input.Type() == kSparseMatrix) {
+    Matrix<BaseFloat> in, out;
```
please put a comment here mentioning that this is very ineffcient,  and if it ever becomes a bottleneck we'll implement something smarter.
Also, 'out' doesn't need to be redeclared.  And I don't like the variables 'in' and 'input' existing, since the two words are synonyms.  I'd prefer in_mat vs in.

danpovey commented on src/util/kaldi-holder.cc:
```diff
@@ -37,41 +39,104 @@ bool ExtractObjectRange(const Matrix<Real> &input, const std::string &range,
+    Matrix<BaseFloat> in, out;
+    input.GetMatrix(&in);
+    ExtractObjectRange(in, range, &out);
+  } else {
```
I'd remove this else, make the previous 'else if' the else, and assert there that input.Type() == kSparseMatrix.
There are no other types that exist in the code.

danpovey commented on src/util/kaldi-holder.cc:
```diff
@@ -37,41 +39,104 @@ bool ExtractObjectRange(const Matrix<Real> &input, const std::string &range,
+bool ExtractObjectRange(const GeneralMatrix &input, const std::string &range,
+                        GeneralMatrix *output) {
+
+  Matrix<BaseFloat> out;
```
maybe call this out_mat, and rename 'output' to 'out'?  I don't like having two variables that only differ in how they are abbreviated (out vs output).

danpovey commented on src/util/kaldi-holder.cc:
```diff
@@ -37,41 +39,104 @@ bool ExtractObjectRange(const Matrix<Real> &input, const std::string &range,
+                   - row_range[0] + 1,
+        col_size = col_range[1] - col_range[0] + 1;
+  /*
+  KALDI_WARN << "size: " << input.NumRows() << "x" << input.NumCols();
```
please remove this comment.

danpovey commented on src/util/kaldi-holder.cc:
```diff
@@ -37,41 +39,104 @@ bool ExtractObjectRange(const Matrix<Real> &input, const std::string &range,
+                        Matrix<Real> *output) {
+  std::vector<int32> row_range, col_range;
+
+  if(!ParseMatrixRangeSpecifier(range, input.NumRows(), input.NumCols(),
```
space after if.

jtrmal commented:
I addressed the comments + tested by decoding of the previously failing task.
Using this I got the same result as I got using the previous quickfix (copy-feats | apply-cmvn)
